Spread operator causes Typescript ParserError

# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Using the new PostgREST spread operator works, but leads to a Typescript error:

`error TS2339: Property 'invitees' does not exist on type 'ParserError<"Expected identifier at `...contact(id,name,email,contact_user_id))`">`

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Add spread (`...`) to a `select()`.
2. Compile with `tsc`

## Expected behavior

Types should successfully be generated when queries include the spread operator.

## System information

- Version of supabase-js: 2.32.0
- Version of Node.js: Cloudflare Workers

--- a/src/select-query-parser.ts
+++ b/src/select-query-parser.ts
@@ -191,6 +191,17 @@ type ConstructFieldDefinition<
   Field
 > = Field extends { star: true }
   ? Row
+  : Field extends { spread: true; original: string; children: unknown[] }
+  ? GetResultHelper<
+      Schema,
+      (Schema['Tables'] & Schema['Views'])[Field['original']]['Row'],
+      Field['original'],
+      (Schema['Tables'] & Schema['Views'])[Field['original']] extends { Relationships: infer R }
+        ? R
+        : unknown,
+      Field['children'],
+      unknown
+    >
   : Field extends { name: string; original: string; hint: string; children: unknown[] }
   ? {
       [_ in Field['name']]: GetResultHelper<
@@ -394,12 +405,20 @@ type ParseField<Input extends string> = Input extends ''
  * - `*`
  * - a field, as defined above
  * - a renamed field, `renamed_field:field`
+ * - a spread field, `...field`
  */
 type ParseNode<Input extends string> = Input extends ''
   ? ParserError<'Empty string'>
   : // `*`
   Input extends `*${infer Remainder}`
   ? [{ star: true }, EatWhitespace<Remainder>]
+  : // `...field`
+  Input extends `...${infer Remainder}`
+  ? ParseField<EatWhitespace<Remainder>> extends [infer Field, `${infer Remainder}`]
+    ? Field extends { children: unknown[] }
+      ? [Prettify<{ spread: true } & Field>, EatWhitespace<Remainder>]
+      : ParserError<'Unable to parse spread resource'>
+    : ParserError<'Unable to parse spread resource'>
   : ParseIdentifier<Input> extends [infer Name, `${infer Remainder}`]
   ? EatWhitespace<Remainder> extends `::${infer _Remainder}`
     ? // `field::`

--- a/test/index.test-d.ts
+++ b/test/index.test-d.ts
@@ -1,6 +1,7 @@
 import { expectError, expectType } from 'tsd'
 import { PostgrestClient, PostgrestSingleResponse } from '../src/index'
 import { SelectQueryError } from '../src/select-query-parser'
+import { Prettify } from '../src/types'
 import { Database, Json } from './types'
 
 const REST_URL = 'http://localhost:3000'
@@ -51,6 +52,31 @@ const postgrest = new PostgrestClient<Database>(REST_URL)
 // cannot update non-updatable columns
 {
   expectError(postgrest.from('updatable_view').update({ non_updatable_column: 0 }))
+}
+
+// spread resource with single column in select query
+{
+  const { data, error } = await postgrest
+    .from('messages')
+    .select('message, ...users(status)')
+    .single()
+  if (error) {
+    throw new Error(error.message)
+  }
+  expectType<{ message: string | null; status: Database['public']['Enums']['user_status'] | null }>(
+    data
+  )
+}
+
+// spread resource with all columns in select query
+{
+  const { data, error } = await postgrest.from('messages').select('message, ...users(*)').single()
+  if (error) {
+    throw new Error(error.message)
+  }
+  expectType<Prettify<{ message: string | null } & Database['public']['Tables']['users']['Row']>>(
+    data
+  )
 }
 
 // json accessor in select query

